src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -918,19 +860,188 @@ export class SearchSource {
+    return [...new Set([...bodyFieldNames, ...fieldsFromSource])];
+  }
+
+  private filterScriptFields({
+    uniqFieldNames,
+    scriptFields,
+  }: {
+    uniqFieldNames: any;
+    scriptFields: any;
+  }) {
+    return uniqFieldNames.includes('*')
+      ? scriptFields
+      : // filter down script_fields to only include items specified
+        pick(
+          scriptFields,
+          Object.keys(scriptFields).filter((f) => uniqFieldNames.includes(f))
+        );
+  }
+
+  private getBuiltEsQuery({
+    index,
+    query,
+    filters,
+    getConfig,
+    sort,
+  }: {
+    index: any;
+    query: any;
+    filters: any;
+    getConfig: (key: string) => any;
+    sort?: any;
+  }) {
+    // If sorting by _score, build queries in the "must" clause instead of "filter" clause to enable scoring
+    const filtersInMustClause = (sort ?? []).some((srt: EsQuerySortValue[]) =>
+      srt.hasOwnProperty('_score')
+    );
+    const esQueryConfigs = {
+      ...getEsQueryConfig({ get: getConfig }),
+      filtersInMustClause,
     };
+    return buildEsQuery(index, query, filters, esQueryConfigs);
+  }
+
+  private getRemainingFields({
+    uniqFieldNames,
+    scriptFields,
+    runtimeFields,
+    _source,
+  }: {
+    uniqFieldNames: any;
+    scriptFields: any;
+    runtimeFields: any;
+    _source: any;
+  }) {
+    return difference(uniqFieldNames, [
+      ...Object.keys(scriptFields),
+      ...Object.keys(runtimeFields),
+    ]).filter((remainingField) => {
+      if (!remainingField) return false;
+      if (!_source || !_source.excludes) return true;
+      return !_source.excludes.includes(remainingField);
+    });
+  }
+
+  private getFieldsList({
+    index,
+    fields,
+    docvalueFields,
+    fieldsFromSource,
+    filteredDocvalueFields,
+    metaFields,
+    fieldListProvided,
+    sourceFieldsProvided,
+  }: {
+    index?: DataView;
+    fields: any;
+    docvalueFields: any;
+    fieldsFromSource: any;
+    filteredDocvalueFields: any;
+    metaFields: string;
+    fieldListProvided: boolean;
+    sourceFieldsProvided: boolean;
+  }) {
+    if (fieldListProvided || sourceFieldsProvided) {
+      // if items that are in the docvalueFields are provided, we should
+      // make sure those are added to the fields API unless they are
+      // already set in docvalue_fields
+      return sourceFieldsProvided
+        ? [
+            ...fields,
+            ...filteredDocvalueFields.filter((fld: SearchFieldValue) => {
+              return (
+                fieldsFromSource.includes(this.getFieldName(fld)) &&
+                !(docvalueFields || [])
+                  .map((d: string | Record<string, SearchFieldValue>) => this.getFieldName(d))
+                  .includes(this.getFieldName(fld))
+              );
+            }),
+          ]
+        : this.getUniqueFields({
+            index,
+            fields,
+            metaFields,
+            filteredDocvalueFields,
+          });
```

Comment:
nit: maybe we can rewrite it in a cleaner way?

```suggestion
      if(!sourceFieldsProvided){
        return this.getUniqueFields({
            index,
            fields,
            metaFields,
            filteredDocvalueFields,
          });
      }
      return [
            ...fields,
            ...filteredDocvalueFields.filter((fld: SearchFieldValue) => {
              const fldName = this.getFieldName(fld);
              return (
                fieldsFromSource.includes(fldName) &&
                !(docvalueFields || [])
                  .map((d: string | Record<string, SearchFieldValue>) => this.getFieldName(d)).includes(fldName)
              );
            }),
          ];
```

I would also put `docvalueFields` into a `Set` to avoid too many iterations (an `.includes` over any `docvalueFields` item), but I am unsure about the cardinality here.